src/daft-micropartition/src/micropartition.rs

```diff
@@ -178,9 +177,82 @@ fn materialize_scan_task(
+                        crate::python::read_parquet_into_py_table(
+                            py,
+                            url,
+                            cast_to_schema.clone().into(),
+                            (*coerce_int96_timestamp_unit).into(),
+                            scan_task.storage_config.clone().into(),
+                            scan_task.columns.as_ref().map(|cols| cols.as_ref().clone()),
+                            scan_task.limit,
+                        )
+                        .map(|t| t.into())
+                        .context(PyIOSnafu)
+                    })
+                    .collect::<crate::Result<Vec<_>>>()
+                })?,
+                FileFormatConfig::Csv(CsvSourceConfig {
+                    has_headers,
+                    delimiter,
+                    double_quote,
+                    ..
+                }) => Python::with_gil(|py| {
+                    urls.map(|url| {
+                        crate::python::read_csv_into_py_table(
+                            py,
+                            url,
+                            *has_headers,
+                            delimiter,
+                            *double_quote,
+                            cast_to_schema.clone().into(),
+                            scan_task.storage_config.clone().into(),
+                            scan_task.columns.as_ref().map(|cols| cols.as_ref().clone()),
+                            scan_task.limit,
+                        )
+                        .map(|t| t.into())
+                        .context(PyIOSnafu)
+                    })
+                    .collect::<crate::Result<Vec<_>>>()
+                })?,
+                FileFormatConfig::Json(_) => Python::with_gil(|py| {
+                    urls.map(|url| {
+                        crate::python::read_json_into_py_table(
+                            py,
+                            url,
+                            cast_to_schema.clone().into(),
+                            scan_task.storage_config.clone().into(),
+                            scan_task.columns.as_ref().map(|cols| cols.as_ref().clone()),
+                            scan_task.limit,
+                        )
+                        .map(|t| t.into())
+                        .context(PyIOSnafu)
+                    })
+                    .collect::<crate::Result<Vec<_>>>()
+                })?,
+            }
         }
     };
+    let cast_to_schema = match &column_names {
+        None => Ok(cast_to_schema),
+        Some(column_names) => Ok(Arc::new(
+            Schema::new(
+                cast_to_schema
+                    .names()
+                    .iter()
+                    .filter(|name| column_names.contains(&name.as_str()))
+                    .map(|name| cast_to_schema.get_field(name).unwrap().clone())
+                    .collect(),
+            )
+            .context(DaftCoreComputeSnafu)?,
+        )),
+    }?;
```

Comment:
@jaychia I moved the field pruning on the schema after the read, since I believe that all of the I/O layers expect the schema to give the full (pre-projection) schema of the file, rather than the pruned schema. I at least know that this is the case for the CSV reader.

Reply:
> I believe that all of the I/O layers expect the schema to give the full (pre-projection) schema of the file

I'm not sure if we can make this assumption actually, since I believe we currently have a mechanism for users to "provide a schema" (`schema_hints`), which could already be a pruned subset of each files' schema.

I.e. something like:

```
daft.read_csv("files/**.csv", schema_hints={"foo": daft.DataType.int64()})
```

I think if a user ran the above, we would expect it to work on all the CSV files with the assumption that each CSV file could have any number of columns, but they should all at least have one column called `"foo"` that can be casted to int64.

Lmk if this makes sense?

Reply:
Offline discussion:

1. We should let `read_csv` do its thing and not assume anything about schema ordering
2. Schema/column ordering is enforced later on using `.cast_to_schema`
3. `read_csv` doesn't take a schema as an argument, and will perform its own schema inference